### ISSUE TITLE
Mass Upload Logs Feature

### ISSUE THREAD

**Issue by ocbaker** [exceptionless/Exceptionless#303]
Hey,

TLDR:

I'd like to see:

- The ability to customise servers so you can upload events from anytime in the past, not just from the last 3 days.
- For the Exceptionless UI to be more clear about when it is adding restrictions to your filter and what those restrictions are.
- For it to be easier to mass upload events.
- The ability to bulk edit existing events in Exceptionless, especially for things like combining stacks of similar events.

---
Story:

With trying to introduce Exceptionless into my current workplace to replace our Textfile based logging I've found that the lack of ability to upload in bulk or upload truly historical data is a big hurdle towards convincing managers of the benefits of using a system like exceptionless over current methods, especially as it isn't always easy to drag and drop tools like Exceptionless into legacy codebases as a spike.

Currently what I have set up is a Self-Hosted Exceptionless instance that I feed existing logs into using a tool I've created which parses a large bundle of logs copied from production (About 500MB from a production client who is having issues) and it parses the logs into .NET objects and then feeds each log message as a Log event into the Exceptionless Server.

However there are a few issues with Exceptionless that hinder this.

1, You can't submit events to the server older than 3-4 days. (So uploading these historical logs is general done using the time of the log message and the current Day-1 so logging events get correctly displayed in the graph)

2, After editing the server source to allow me to upload older log events I've noticed that the queries for events that the Exceptionless UI sends back to Elastic has a built in query element restricting the "from" datetime. But there is no way to configure this and more importantly there is no way for the user to know that even though in the UI they're filtering for events from 2 months ago, Exceptionless has not really quired for events from that daterange. (If you query Elastic using Kibana you'll get the full results from any specified daterange, even when Exceptionless UI comes up blank)

3, There is no way to bulk edit existing events sent to exceptionless. Especially when uploading historical data you notice that sometimes you've not set the source perfectly for a certain variant of message and that has caused a lot of effectively duplicate messages (with slightly different variables in them) to be counted as unique. Currently the only way to rectify this is to wipe all already uploaded events and edit my parsing tool so that when I resend all my events it handles things better when setting event information before sending to the server. Personally I feel that it would be much nicer if you could select a message in the UI and set a new Rule (Like you might in Outlook) where you can define something like: "If a source comes in with text matching this regex, change source message to X and add additional values A,B,C"

---

Even with the rudimentary tooling I've created to convert and uploading some of our existing production logs already people I've shown it to are very impressed and want to see more. I really feel that it would be very beneficial for Exceptionless to build their own tooling which can parse logs and upload them in bulk and to allow uploading these logs as historical events.

Tooling like this can be advertised to potential customers who might be interested but are hesitant of the added work of properly migrating their existing codebases. If larger, older codebases can easily get off the ground running by being able to upload their existing log output (Provided they at least produce formattable logs) then they'll be able to get some of the benefits of Exceptionless earlier without having to stress about turning around their whole codebase to see some payback.

**Comment by ejsmith** [comment]
Thanks for the feedback!

1. We disabled going back as far as you want in our code because it has a lot of performance implications, but I think we should probably make that configurable and let self-hosters choose their own setting.

2. If you change the date filter then you should be able to view events as far back as you want. If you are getting restricted to 3 days then the system must think that you are on a free plan. The default plan for self-hosting should be unlimited. Not sure what is happening there.

3. I would really like to have bulk editing as well. Ideally though, we want events themselves to be immutable because it provides a ton of perf benefits. But we should be able to bulk edit the stacks to change things like tags and other properties.

Would love to get help with these items. Do you want to take a stab at creating a pull request for some of these things?

**Comment by niemyjski** [comment]
@ocbaker would you mind submitting a pr for any of these?

**Comment by niemyjski** [comment]
I'm going to close this due to the previous reply by @ejsmith. I think it's a valid feature request to have bulk editing the stacks. Would you mind submitting a new issue or a pr (even better :)) for this.
